File: 프로그래머스/lv3/43238. 입국심사/입국심사.py
# 임의의 시간을 구하고, 그 시간동안 몇명이 심사를 받을 수 있는 지 확인하며 이분탐색한다.

def solution(n, times):
    answer = 0
    left = min(times)
    right = max(times) * n
    
    while left <= right:
        mid = (left + right) // 2
        people = 0
        for time in times:
            people += mid // time
            
        if people >= n:
            answer = mid
            right = mid - 1
            
        else:
            left = mid + 1
        
    return answer

File: 프로그래머스/3/43238. 입국심사/입국심사.js
function solution(n, times) {
    let answer = 0;
    let start = 1;
    let end = times[times.length -1] * n; // 가장 오래걸리는 시간
    
    while (start <= end) {
        let mid = Math.floor((start + end) / 2); // 임의의 시간을 계산
        let people = times.reduce((acc, cur) => acc + Math.floor(mid / cur), 0) 
        // 임의의 시간 mid 동안 심사관 한 명당 심사할 수 있는 인원을 계산
        
        if (people < n) {
            start = mid + 1;
        } else {
            end = mid - 1;
            answer = mid
        }
    }
    return answer;
}